### PR TITLE
support callback for xtoken receive or rollback

### PR DESCRIPTION
**Pull request by xiaoch05**

### CODE REVIEW

File: helix-contract/contracts/interfaces/IUniswapV3SwapCallback.sol [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    /// @notice Called to `msg.sender` after executing a swap via IUniswapV3Pool#swap.
+    /// @dev In the implementation you must pay the pool tokens owed for the swap.
+    /// The caller of this method must be checked to be a UniswapV3Pool deployed by the canonical UniswapV3Factory.
+    /// amount0Delta and amount1Delta can both be 0 if no tokens were swapped.
+    /// @param amount0Delta The amount of token0 that was sent (negative) or must be received (positive) by the pool by
+    /// the end of the swap. If positive, the callback must send that amount of token0 to the pool.
+    /// @param amount1Delta The amount of token1 that was sent (negative) or must be received (positive) by the pool by
+    /// the end of the swap. If positive, the callback must send that amount of token1 to the pool.
+    /// @param data Any data passed through by the caller via the IUniswapV3PoolActions#swap call
+    function uniswapV3SwapCallback(
```
Comment (hujw77) [inline helix-contract/contracts/interfaces/IUniswapV3SwapCallback.sol:16]: rename to `helixCallback`

File: helix-contract/contracts/mapping-token/v3/base/xTokenIssuing.sol [hunk 2]
```diff
@@ -113,9 +125,14 @@ contract xTokenIssuing is xTokenBridgeBase {
...
-            IGuard(_guard).deposit(uint256(transferId), xToken, _recipient, _amount);
+            IGuardV3(_guard).deposit(uint256(transferId), xToken, _recipient, _amount, _extData);
         } else {
             xTokenErc20(xToken).mint(_recipient, _amount);
+            if (ERC165Checker.supportsInterface(_recipient, type(IUniswapV3SwapCallback).interfaceId)) {
+                CallbackInfo memory callbackInfo = CallbackInfo(msg.sig, transferId, xToken, _extData);
+                bytes memory data = abi.encode(callbackInfo);
+                IUniswapV3SwapCallback(_recipient).uniswapV3SwapCallback(int256(_amount), int256(_amount), data);
+            }
         }
```
Comment (xiaoch05) [inline helix-contract/contracts/mapping-token/v3/base/xTokenIssuing.sol:136]: guard is a special callback contract

File: helix-contract/contracts/mapping-token/v3/base/xTokenIssuing.sol [hunk 3]
```diff
@@ -230,6 +251,11 @@ contract xTokenIssuing is xTokenBridgeBase {
         require(xToken != address(0), "xToken not exist");
 
         xTokenErc20(xToken).mint(_originalSender, _amount);
+        if (ERC165Checker.supportsInterface(_originalSender, type(IUniswapV3SwapCallback).interfaceId)) {
+            CallbackInfo memory callbackInfo = CallbackInfo(msg.sig, transferId, _originalToken, "");
+            bytes memory data = abi.encode(callbackInfo);
+            IUniswapV3SwapCallback(_originalSender).uniswapV3SwapCallback(int256(_amount), int256(_amount), data);
+        }
```
Comment (xiaoch05) [inline helix-contract/contracts/mapping-token/v3/base/xTokenIssuing.sol:258]: use different interfaces.

**Comment by xiaoch05** [comment]
mapping-token => xToken

File: helix-contract/contracts/mapping-token/v3/interfaces/IxTokenBacking.sol [hunk 4]
```diff
@@ -8,7 +8,8 @@ interface IxTokenBacking {
         address originalSender,
         address recipient,
         uint256 amount,
-        uint256 nonce
+        uint256 nonce,
+        bytes calldata extData
     ) external;
```
Comment (xiaoch05) [inline helix-contract/contracts/mapping-token/v3/interfaces/IxTokenBacking.sol:14]: xUnlock

File: helix-contract/contracts/mapping-token/v3/interfaces/IxTokenBacking.sol [hunk 5]
```diff
@@ -8,7 +8,8 @@ interface IxTokenBacking {
...
 
     function handleUnlockForIssuingFailureFromRemote(
```
Comment (xiaoch05) [inline helix-contract/contracts/mapping-token/v3/interfaces/IxTokenBacking.sol:15]: xRollbackLockAndRemoteIssuing

Comment (xiaoch05) [reply helix-contract/contracts/mapping-token/v3/interfaces/IxTokenBacking.sol:15]: xRollbackLock

Comment (xiaoch05) [reply helix-contract/contracts/mapping-token/v3/interfaces/IxTokenBacking.sol:15]: xRollbackLockForIssuingFailure

File: helix-contract/contracts/interfaces/IUniswapV3SwapCallback.sol [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity >=0.8.17;
+
+/// @title Callback for IUniswapV3PoolActions#swap
+/// @notice Any contract that calls IUniswapV3PoolActions#swap must implement this interface
+interface IUniswapV3SwapCallback {
```
Comment (xiaoch05) [inline helix-contract/contracts/interfaces/IUniswapV3SwapCallback.sol:6]: xTokenCallback

File: helix-contract/contracts/interfaces/IUniswapV3SwapCallback.sol (hunk 1 above)
Comment (xiaoch05) [reply helix-contract/contracts/interfaces/IUniswapV3SwapCallback.sol:16]: xTokenCallbackA/B

File: helix-contract/contracts/mapping-token/v3/base/xTokenBacking.sol [hunk 7]
```diff
@@ -120,53 +126,65 @@ contract xTokenBacking is xTokenBridgeBase {
...
+                bytes memory data = abi.encode(callbackInfo);
+                IUniswapV3SwapCallback(_recipient).uniswapV3SwapCallback(int256(_amount), int256(_amount), data);
+            }
         } else {
             uint allowance = IERC20(_token).allowance(address(this), _guard);
             require(IERC20(_token).approve(_guard, allowance + _amount), "Backing:approve token transfer to guard failed");
-            IGuard(_guard).deposit(uint256(_transferId), _token, _recipient, _amount);
+            IGuardV3(_guard).deposit(uint256(_transferId), _token, _recipient, _amount, _extData);
         }
     }
```
Comment (xiaoch05) [inline helix-contract/contracts/mapping-token/v3/base/xTokenBacking.sol:190]: is it possiable to use the same callback for native and erc20 token

File: helix-contract/contracts/mapping-token/v3/base/xTokenBridgeBase.sol [hunk 8]
```diff
@@ -25,6 +25,13 @@ contract xTokenBridgeBase is Initializable, Pausable, AccessController, DailyLim
         bool hasRefundForFailed;
     }
 
+    struct CallbackInfo {
+        bytes4 sig;
+        bytes32 transferId;
+        address token;
+        bytes extData;
+    }
```
Comment (xiaoch05) [inline helix-contract/contracts/mapping-token/v3/base/xTokenBridgeBase.sol:33]: remove this struct

**Comment by xiaoch05** [comment]
change native token to wtoken

**Pull request merged**